### PR TITLE
turn Timeline::layers into tokio::sync::RwLock

### PR DESCRIPTION
**Pull request by problame**
This is preliminary work for/from #4220 (async `Layer::get_value_reconstruct_data`).

# Full Stack Of Preliminary PRs

Thanks to the countless preliminary PRs, this conversion is relatively straight-forward.

1. Clean-ups
  * https://github.com/neondatabase/neon/pull/4316
  * https://github.com/neondatabase/neon/pull/4317
  * https://github.com/neondatabase/neon/pull/4318
  * https://github.com/neondatabase/neon/pull/4319
  * https://github.com/neondatabase/neon/pull/4321
  * Note: these were mostly to find an alternative to #4291, which I thought we'd need in my original plan where we would need to convert `Tenant::timelines` into an async locking primitive (#4333). In reviews, we walked away from that, but these cleanups were still quite useful.
2. https://github.com/neondatabase/neon/pull/4364
3. https://github.com/neondatabase/neon/pull/4472
4. https://github.com/neondatabase/neon/pull/4476
5. https://github.com/neondatabase/neon/pull/4477
6. https://github.com/neondatabase/neon/pull/4485

# Significant Changes In This PR

## `compact_level0_phase1` & `create_delta_layer`

This commit partially reverts

   "pgserver: spawn_blocking in compaction (#4265)"
    4e359db4c78e0fd3d3e8f6a69baac4fb5b80b752.

Specifically, it reverts the `spawn_blocking`-ificiation of `compact_level0_phase1`.
If we didn't revert it, we'd have to use `Timeline::layers.blocking_read()` inside `compact_level0_phase1`.
That would use up a thread in the `spawn_blocking` thread pool, which is hard-capped.

I considered wrapping the code that follows the second `layers.read().await`
into `spawn_blocking`, but there are lifetime issues with `deltas_to_compact`.

Also, this PR switches the `create_delta_layer` _function_ back to async, and uses `spawn_blocking` inside to run the code that does sync IO, while keeping the code
that needs to lock `Timeline::layers` async.

## `LayerIter` and `LayerKeyIter` `Send` bounds

I had to add a `Send` bound on the `dyn` type that `LayerIter` and `LayerKeyIter` wrap.
Why? Because we now have the second `layers.read().await` inside `compact_level0_phase`,
and these iterator instances are held across that await-point.

More background: https://github.com/neondatabase/neon/pull/4462#issuecomment-1587376960

## `DatadirModification::flush`

Needed to replace the `HashMap::retain` with a hand-rolled variant because `TimelineWriter::put` is now async.

### CODE REVIEW

**Comment by problame** [comment]
Rebased [onto latest base](https://github.com/neondatabase/neon/pull/4364#issuecomment-1581101148)

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -1133,15 +1133,19 @@ impl<'a> DatadirModification<'a> {
...
-                false
-            } else {
-                true
+        let mut retain = HashSet::new();
+        let result = async {
+            for (key, value) in &self.pending_updates {
+                if is_rel_block_key(*key) || is_slru_block_key(*key) {
+                    writer.put(*key, self.lsn, value).await?;
+                } else {
+                    retain.insert(*key);
```
Comment (skyzh) [inline pageserver/src/pgdatadir_mapping.rs:1142]: I don't think this is equivalent to the original one. If we have a sequence key1, key2, key3, whereas key2 fails on put, the original code will retain key3, while this one will remove everything after the first error.

**Review by skyzh — COMMENTED** [review]
Rest LGTM, good work!

File: pageserver/src/pgdatadir_mapping.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/pgdatadir_mapping.rs:1142]: 😬 good catch, I implemented that in a hurry. Curious to see whether tests will fail.

Comment (problame) [reply pageserver/src/pgdatadir_mapping.rs]: 3fbe502a91a80659e214735feaf97644aebe3770

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -3006,34 +2997,49 @@ impl Timeline {
...
+                // them all in parallel.
+
+                // First sync the delta layer. We still use par_fsync here to keep everything consistent. Feel free to replace
+                // this with a single fsync in future refactors.
+                par_fsync::par_fsync(&[new_delta_path.clone()]).context("fsync of delta layer")?;
+                // Then sync the parent directory.
+                par_fsync::par_fsync(&[self_clone
+                    .conf
+                    .timeline_path(&self_clone.timeline_id, &self_clone.tenant_id)])
+                .context("fsync of timeline dir")?;
```
Comment (hlinnaka) [inline pageserver/src/tenant/timeline.rs:3031]: we can use `par_fsync_async` if we move this outside the task_in_place block.

File: pageserver/src/tenant/timeline.rs [hunk 3]
```diff
@@ -3254,19 +3262,15 @@ impl Timeline {
...
             .collect::<Vec<_>>();
-
-        par_fsync::par_fsync_async(&all_paths)
-            .await
-            .context("fsync of newly created layer files")?;
-
-        par_fsync::par_fsync_async(&[self.conf.timeline_path(&self.timeline_id, &self.tenant_id)])
-            .await
-            .context("fsync of timeline dir")?;
+        par_fsync::par_fsync(&all_paths).context("fsync of newly created layer files")?;
```
Comment (hlinnaka) [inline pageserver/src/tenant/timeline.rs:3269]: Why this change? We would prefer the async `par_fsync_async` variant in async code, right?

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/tenant/timeline.rs (hunk 3 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:3269]: also this part should not be reverted because we should first fsync image layers, then the timeline directory.

**Review by skyzh — APPROVED** [review]
LGTM except the fsync changes. Some of them should be kept as-is because the compaction `spawn_blocking` PR also fixes some bugs that should not be reverted... 🤪 (See comment at timeline.rs L3269)

File: pageserver/src/tenant/timeline.rs [hunk 4]
```diff
@@ -125,7 +125,7 @@ pub struct Timeline {
 
     pub pg_version: u32,
 
-    pub(super) layers: RwLock<LayerMap<dyn PersistentLayer>>,
+    pub(crate) layers: Arc<tokio::sync::RwLock<LayerMap<dyn PersistentLayer>>>,
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/timeline.rs]: > I guess we need to take OwnedReadGuard somewhere

Correct, I needed that in some earlier version, but, not anymore.

Will remove it

6c4024bda65f271bb80693007cf6ffc833df47dd

**Comment by problame** [comment]
Ok, so, I re-cherry-picked @skyzh 's `pgserver: spawn_blocking in compaction (#4265)` .

To the best of my knowledge, all that it reverts now is the `spawn_blocking`-ingification of `compact_level0_phase1`.

So, we're doing sync IO again inside the async `compact_level0_phase1`.
I'll try to eliminate that by wrapping the writes into `block_in_place` or `spawn_blocking`.
Will push such a commit later.


But what remains is the CPU work of the kmerge.
The lifetimes around that are non-trivial, I think it's better to fix that up / make it `spawn_blocking` in a later PR.

**Comment by problame** [comment]
> I'll try to eliminate that by wrapping the writes into block_in_place or spawn_blocking.
> Will push such a commit later

Scoped this out. I don't think it's a good idea to wrap just the `put_value` call into a `spawn_blocking`. Because we only do IO (or CPU-"intensive" prefix compression) in a small fraction of the overall `put_value` calls.

So, I'm leaning toward just taking the regression here.
Let's quantify the regression.

@skyzh do you have a benchmark / numbers that you ran when you implemented the `spawn_blocking`-ingification?

File: pageserver/src/tenant/timeline.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs]: I did that in https://github.com/neondatabase/neon/pull/4441/commits/95bd6c4b77e1824c7c14b26a47609fe943e45ae3 , please sanity-check

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3031]: Not sure if that's actually better in terms of performance, because, these async variants internally just use spawn_blocking as well.
So we'd just add an additional spawn_blocking call worth of latency.

Now I don't know if the same call can be made for block_in_place. AFAIK it just informs tokio runtime that the current worker thread might block?
So, maybe it's better to keep the block_in_place as short as possible?
In that case, your suggestion makes sense.

Would appreciate comment from someone with more tokio experience

File: pageserver/src/pgdatadir_mapping.rs [hunk 5]
```diff
@@ -1133,15 +1133,17 @@ impl<'a> DatadirModification<'a> {
...
+            if result.is_ok() && (is_rel_block_key(*key) || is_slru_block_key(*key)) {
+                result = writer.put(*key, self.lsn, value).await;
             } else {
-                true
+                retain.insert(*key);
             }
-        });
+        }
+        self.pending_updates.retain(|key, _| retain.contains(key));
         result?;
```
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/pgdatadir_mapping.rs]: Looks ok to me. Ideally it should be drain_filter that is forever unstable....

Looked a bit more, the library that standard hashmap is based on provides [extract_if](https://docs.rs/hashbrown/latest/hashbrown/struct.HashMap.html#method.extract_if) method, which looks suitable here. And we already have dependency on the library in our workspace (directly and indirectly)

I'm ok with either option.

File: pageserver/src/tenant/timeline.rs [hunk 6]
```diff
@@ -3006,34 +2998,48 @@ impl Timeline {
...
-    fn create_delta_layer(
+    async fn create_delta_layer(
         self: &Arc<Self>,
         frozen_layer: &InMemoryLayer,
     ) -> anyhow::Result<(LayerFileName, LayerFileMetadata)> {
-        // Write it out
-        let new_delta = frozen_layer.write_to_disk()?;
-        let new_delta_path = new_delta.path();
-        let new_delta_filename = new_delta.filename();
+        // TODO figure out how to use spawn_blocking. Can't use it because frozen_layer is not 'static
```
Comment (LizardWizzard) [inline pageserver/src/tenant/timeline.rs:3005]: One function up the stack (flush_frozen_layer) it is an `Arc<InMemoryLayer>` so you can clone it the usual way

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs:3031]: I would prefer async version because it uses shared tokio threadpool instead of our own that spawns threads for each call.

File: pageserver/src/tenant/timeline.rs [hunk 7]
```diff
@@ -3006,34 +2998,48 @@ impl Timeline {
...
+                // First sync the delta layer. We still use par_fsync here to keep everything consistent. Feel free to replace
+                // this with a single fsync in future refactors.
+                par_fsync::par_fsync(&[new_delta_path.clone()]).context("fsync of delta layer")?;
+                // Then sync the parent directory.
+                par_fsync::par_fsync(&[self_clone
+                    .conf
+                    .timeline_path(&self_clone.timeline_id, &self_clone.tenant_id)])
+                .context("fsync of timeline dir")?;
+
+                let sz = new_delta_path.metadata()?.len();
```
Comment (LizardWizzard) [inline pageserver/src/tenant/timeline.rs:3033]:
```suggestion
                let sz = new_delta.desc.file_size;
```

I was thinking that it tracks number of bytes written, but it still calls to `metadata()` underneath which shouldnt be needed

File: pageserver/src/tenant/timeline.rs [hunk 8]
```diff
@@ -3006,34 +2998,48 @@ impl Timeline {
...
+                //
+                // We must also fsync the timeline dir to ensure the directory entries for
+                // new layer files are durable
+                //
+                // TODO: If we're running inside 'flush_frozen_layers' and there are multiple
+                // files to flush, it might be better to first write them all, and then fsync
+                // them all in parallel.
+
+                // First sync the delta layer. We still use par_fsync here to keep everything consistent. Feel free to replace
+                // this with a single fsync in future refactors.
```
Comment (LizardWizzard) [inline pageserver/src/tenant/timeline.rs:3025]: > Feel free to replace this with a single fsync in future refactors.

NIT: I know that you didnt change that, but it looks incorrect to me. From the point of consistency it cannot be done this way (which is implied by previous sentence)

File: pageserver/src/tenant/timeline.rs [hunk 9]
```diff
@@ -3671,21 +3679,12 @@ impl Timeline {
...
-        } = tokio::task::spawn_blocking(move || {
-            let _g = span.entered();
-            this.compact_level0_phase1(layer_removal_cs_inner, target_file_size, &ctx_inner)
-        })
-        .await
-        .context("compact_level0_phase1 spawn_blocking")
-        .map_err(CompactionError::Other)
-        .and_then(|res| res)?;
+        } = self
+            .compact_level0_phase1(layer_removal_cs.clone(), target_file_size, ctx)
```
Comment (LizardWizzard) [inline pageserver/src/tenant/timeline.rs:3686]: It still looks quite CPU heavy to me. So I would keep it in `spawn_blocking` and use `layers.blocking_write` to access layer map

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3031]: But we need the spawn_blocking because the writing of the delta layer directly above this piece of code is currently using sync code...

File: pageserver/src/tenant/timeline.rs (hunk 6 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3005]: a4c4d967d76aab1413e3e07a32b149f8d4e2f0e8

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs:3031]: Hm, dont see any problems there, spawn_blocking to write then await on sync that uses spawn blocking as well. It will reuse threads from tokio pool. Though an important detail here is that there is no actual parallelism happening because with `paths.len() == 1` it wont spawn additional threads. So maybe we can just fsync them without `par` for clarity? Then it will be in the same spawn_blocking until `write_to_disk` becomes async, wdyt?

**Comment by skyzh** [comment]
> do you have a benchmark / numbers that you ran when you implemented the spawn_blocking-ingification?

Unluckily no :( We can look at the daily benchmark number after merging this PR.

**Review by skyzh — APPROVED** [review]

File: pageserver/src/tenant/timeline.rs (hunk 9 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3688]: We've only been using `spawn_blocking` for 2 weeks or so.
It's more important to get this merged than the potential perf improvement that we got there).

I'll create a follow-up issue to get this back.

**Review by LizardWizzard — APPROVED** [review]
LGTM

Discussed follow ups verbally

File: pageserver/src/tenant/timeline.rs (hunk 9 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3688]: #4492

File: pageserver/src/pgdatadir_mapping.rs [hunk 10]
```diff
@@ -1133,16 +1133,17 @@ impl<'a> DatadirModification<'a> {
         let writer = self.tline.writer().await;
 
         // Flush relation and  SLRU data blocks, keep metadata.
-        let mut result: anyhow::Result<()> = Ok(());
-        self.pending_updates.retain(|&key, value| {
-            if result.is_ok() && (is_rel_block_key(key) || is_slru_block_key(key)) {
-                result = writer.put(key, self.lsn, value);
-                false
+        let mut retained_pending_updates = HashMap::new();
```
Comment (shanyp) [inline pageserver/src/pgdatadir_mapping.rs:1136]: use reserve to avoid reallocations
